File: node/blockchain.js
const prompt = require("prompt-sync")({ sigint: true });

function log(m) { console.log(m) }

const MINING_REWARD = 10
const GENESIS_TRANSACTION = {'prevHash': '', 'index': 0, 'transactions': []}
const blockchain = [GENESIS_TRANSACTION];
const openTransactions = [];
const participants = new Set();
const owner = 'Mak';

function getChoice() {
    return prompt("Choice: ");
}

function getUserInput() {
    sender = owner;
    recipient = prompt("Who is the recipient? ");
    amount = parseFloat(prompt("Your transaction amount please: "));
    return {'sender': sender, 'recipient': recipient, 'amount': amount};
}

function getLastTransaction() {
    return blockchain[blockchain.length - 1];
}

function hashBlock(block) {
    return JSON.stringify(block);
}

function verifyTransaction(transaction) {
    let senderBalance = calcBalance(transaction['sender']);
    log(`verifyTransaction(): senderBalance ${senderBalance} : txAmount ${transaction['amount']}`);
    return senderBalance >= transaction['amount'];
}

function addTransaction(sender, recipient, amount=1.0) {
    let tx = {'sender': sender, 'recipient': recipient, 'amount': amount};
    if (verifyTransaction(tx)) {
        openTransactions.push(tx);
        participants.add(sender);
        participants.add(recipient);
        return true
    }
    return false
}

function mineBlock() {
    let rewardTx = {'sender': 'MINING', 'recipient': owner, 'amount': MINING_REWARD};
    openTransactions.push(rewardTx);
    let newBlock = {'prevHash': hashBlock(getLastTransaction()), 
                    'index': blockchain.length, 
                    'transactions': JSON.parse(JSON.stringify(openTransactions))};
    blockchain.push(newBlock);
    openTransactions.length = 0;
}

function verifyChain() {
    if (blockchain.length < 2) { return true; }
    for (index in blockchain) {
        if (index == 0) { continue; }
        if (blockchain[index]['prevHash'] != JSON.stringify(blockchain[index - 1])) {
            return false;
        }
    }
    return true;
}

function calcOpenTransactionsToBeSent(participant) {
    let sendAmount = 0;
    openTransactions.forEach(tx => {
        if (tx['sender'] == participant) {
            sendAmount += tx['amount'];
        }
    });
    return sendAmount
}

function calcBalance(participant) {
    let total = -(calcOpenTransactionsToBeSent(participant));
    blockchain.forEach(elem => {
        let transactions = elem['transactions'];
        transactions.forEach(tx => {
            if (tx['sender'] == participant) {
                total -= tx['amount'];
            }
            if (tx['recipient'] == participant) {
                total += tx['amount'];
            }
        });
    });
    return total;
}



function displayMenu() {
    log('Choose an option:');
    log('   a: Add transaction');
    log('   p: Print chain');
    log('   o: Print openTransactions');
    log('  pp: Print participants');
    log('   m: Mine blocks');
    log('   c: Calc Balances');
    log('   h: Hack');
    log('   t: Run Test Func');
    log('   q: Quit');
}

while (true) {
    displayMenu();
    const choice = getChoice();
    if (choice == 'a') {
        const txDetails = getUserInput()
        if (! addTransaction(txDetails['sender'], txDetails['recipient'], txDetails['amount'])) {
            log('Transaction failed: Insufficient funds!')
        };
    } else if (choice == 'p') {
        log(blockchain);
    } else if (choice == 'o') {
        log(openTransactions);
    } else if (choice == 'pp') {
        log('Participants:')
        participants.forEach(elem => {log(`  - ${elem}`)})
    } else if (choice == 'm') {
        mineBlock();
    } else if (choice == 'c') {
        participants.forEach(p => {log(`${p}: ${calcBalance(p)}`)})
    } else if (choice == 'h') {
        blockchain[0] = -1;
    } else if (choice == 'q') {
        break;
    } else {
        continue;
    }

    if (!verifyChain()) {
        log('INVALID BLOCKCHAIN!!!');
        log('Done!');
        log(blockchain);
        break;
    }

}

File: python/blockchain.py
from tkinter import W

genesis_block = {"previous_hash": "", "index": 0, "transactions": []}
blockchain = [genesis_block]
open_transactions = []
owner = "Mak"


def get_last_transaction():
    return blockchain[-1]


def add_transaction(recipient, sender=owner, amount=1.0):
    transaction = {"recipient": recipient, "sender": sender, "amount": amount}
    open_transactions.append(transaction)


def get_transaction():
    sender = owner
    recipient = input("Who is the recipient? ")
    amount = float(input("Your transaction amount: "))
    return recipient, sender, amount


def get_user_choice():
    return input("Your choice: ")


def hash_block(block):
    # print("".join([str(block[k]) for k in block]))
    return "".join([str(block[k]) for k in block])


def mine_block():
    last_block = blockchain[-1]
    last_block_hash = hash_block(last_block)
    # print(f"Last block hash: {last_block_hash}")
    new_block = {
        "previous_hash": last_block_hash,
        "index": len(blockchain),
        "transactions": open_transactions,
    }
    blockchain.append(new_block)


def print_blockchain_elements():
    output = ""
    # output = ", ".join(map(str, blockchain))
    # output += "\n" + ("-" * 20)
    for block in blockchain:
        output += f"\n{block}"
    print(output)


def verify_chain():
    for i, block in enumerate(blockchain):
        if i > 0:
            if block["previous_hash"] != hash_block(blockchain[i - 1]):
                return False
    return True


def main():
    wait_for_input = True

    while wait_for_input:
        print("Choose Option:")
        print("  a: Add transaction amount")
        print("  m: Mine block")
        print("  p: Print chain")
        print("  h: Hack chain")
        print("  q: Quit")
        choice = get_user_choice()
        if choice == "a":
            recipient, sender, amount = get_transaction()
            add_transaction(recipient, sender, amount)
            print(open_transactions)
        elif choice == "m":
            mine_block()
        elif choice == "p":
            print_blockchain_elements()
        elif choice == "h":
            if len(blockchain) >= 1:
                blockchain[0]["transactions"] = [-1]
        elif choice == "q":
            break
        else:
            continue

        if not verify_chain():
            print("INVALID CHAIN!!!\nDone!")
            print_blockchain_elements()
            wait_for_input = False


if __name__ == "__main__":
    main()
